stream/stream.js

```diff
@@ -2,161 +2,136 @@
-}
-function updateDependency(stream, mustSync) {
-	var state = stream._state, parents = state.parents
-	if (parents.length > 0 && parents.every(active) && (mustSync || parents.some(changed))) {
-		var value = stream._state.derive()
-		if (value === HALT) return unregisterStream(stream)
-		updateState(stream, value)
+			return value
+		})
+		return end
 	}
-}
-function finalize(stream) {
-	stream._state.changed = false
-	for (var id in stream._state.deps) stream._state.deps[id]._state.changed = false
-}
 
-function combine(fn, streams) {
-	if (!streams.every(valid)) throw new Error("Ensure that each item passed to stream.combine/stream.merge is a stream")
-	return initDependency(createStream(), streams, function() {
-		return fn.apply(this, streams.concat([streams.filter(changed)]))
-	})
-}
+	stream.toJSON = function() { return value != null && typeof value.toJSON === "function" ? value.toJSON() : value }
 
-function initDependency(dep, streams, derive) {
-	var state = dep._state
-	state.derive = derive
-	state.parents = streams.filter(notEnded)
+	stream["fantasy-land/map"] = stream.map
+	stream["fantasy-land/ap"] = function(x) { return combine(function(s1, s2) { return s1()(s2()) }, [x, stream]) }
+	stream["fantasy-land/of"] = Stream
 
-	registerDependency(dep, state.parents)
-	updateDependency(dep, true)
+	Object.defineProperty(stream, "end", {
+		get: function() { return end || createEnd() }
+	})
 
-	return dep
-}
-function registerDependency(stream, parents) {
-	for (var i = 0; i < parents.length; i++) {
-		parents[i]._state.deps[stream._state.id] = stream
-		registerDependency(stream, parents[i]._state.parents)
-	}
-}
-function unregisterStream(stream) {
-	for (var i = 0; i < stream._state.parents.length; i++) {
-		var parent = stream._state.parents[i]
-		delete parent._state.deps[stream._state.id]
-	}
-	for (var id in stream._state.deps) {
-		var dependent = stream._state.deps[id]
-		var index = dependent._state.parents.indexOf(stream)
-		if (index > -1) dependent._state.parents.splice(index, 1)
-	}
-	stream._state.state = 2 //ended
-	stream._state.deps = {}
+	return stream
 }
 
-function map(fn) {return combine(function(stream) {return fn(stream())}, [this])}
-function ap(stream) {return combine(function(s1, s2) {return s1()(s2())}, [stream, this])}
-function toJSON() {return this._state.value != null && typeof this._state.value.toJSON === "function" ? this._state.value.toJSON() : this._state.value}
+function combine(fn, streams) {
+	var ready = streams.every(function(s) {
+		if (s["fantasy-land/of"] !== Stream)
+			throw new Error("Ensure that each item passed to stream.combine/stream.merge is a stream")
+		return s.state === "active"
+	})
+	var stream = ready
+		? Stream(fn.apply(null, streams.concat([streams])))
+		: Stream()
+
+	let changed = []
+
+	streams.forEach(function(s) {
+		s.map(function(value) {
+			changed.push(s)
+			if (streams.every(function(s) { return s.state === "active" })) {
```

Comment:
How about doing

`if (ready || (ready =...`?

Reply:
Cool - That's a neat little optimization :) I haven't seen assignments in `ifs predicate` anywhere else in the codebase (correct me if I'm wrong), so I'll probably move it inside..

Reply:
As long as the linter doesn't complain, I don't think it matters much, as long as it works.

Reply:
It also highlights another issue which I don't think we handle right / test for.. If one of the combined streams are ended the combined stream will not be changed even though one of the other ones that aren't ended change.. I think the correct behavior would be for the combined stream to change on any dependent stream change, so I'll change the check to be `s.state !== 'pending'` too, and add a test...

Reply:
Does it currently end all other streams if any stream ends? If so, that's perfectly rational behavior, and arguably makes more sense (it's what RxJS's similar `combineLatest` operator does).

Reply:
No currently it just ends the combined stream. 

I would find it highly odd if it could end any of the other streams being combined. That suddenly changes combine from being a read-only of the combined streams to also control them.
I imagine you'd quickly end up wondering why your stream is ended because it was combined with another stream that ended in a completely separate place..

Reply:
Edit: that's what I meant...I meant ending just the combined stream. Brain lapse here... 😄

Reply:
It's late here, and my brain is slightly fried...

Reply:
Ah ok - good :D hehe.. 

Well, I'm still thinking it would make most sense if it doesn't end until all have ended.. I'll just check up on what flyd does...

Reply:
Flyd continues until all streams are ended, which the current mithril implementation also does (just didn't have a test)..

Reply:
Not sure following flyd blindly is the best move here.

in flyd terms ending a stream means it won't forward values to it's dependents.

e.g.

```js
s = stream(1)
s.end(true)
s() // 1;
s(2)
s() // 2
```

This has the unfortunate side effect that:

```js
s = stream(1)
s2 = stream(2)
s3 = combine((a,b)=> a() + b(), [s, s2]);

s3() // 3
s.end(true);
s(30)
s3() // 32
```

Which is not what many would expect.

I would propose mithril streams go another way, and say that an ended stream is a stream that doesn't accept new values.

Reply:
That's also what I am proposing. My point is that a single ended stream in a combined stream shouldn't prevent the other streams from updating the combined stream?

Reply:
I don't want the behavior in your last example, but rather 
```
s = stream(1)
s2 = stream(2)
s3 = combine((a,b)=> a() + b(), [s, s2]);

s3() // 3
s.end(true);
s2(4)
s(5)
s3() // 5
```

Reply:
We're in agreement there :+1:

Reply:
so in the simplest case a test that does:

```
s = stream(1)
s.end(true)
s(2)

assertEquals(s(), 1)
```